File: Download PDFs AT e SS/DownloaderSS.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using SmartFormat;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Download_PDFs_AT_e_SS
{
    partial class Downloader
    {
        internal static void DownloadExtratoRemuneracoes(int ano, int mes)
        {
            driver.Navigate().GoToUrl("https://app.seg-social.pt/ptss/gr/pesquisa/consultarDR?dswid=7064&frawMenu=1");

            //Coloca as datas nos campos
            string anoMesStr = ano + "-" + (mes < 10 ? ("0" + mes) : mes.ToString());
            string anoMesDiaStr = ano + "-" + (mes < 10 ? ("0" + mes) : mes.ToString()) + "-01";
            ((IJavaScriptExecutor)driver).ExecuteScript("document.getElementById(\"dadosPesquisaDeclaracoes:dataReferenciaFimMonthPicker:calendar_input\").value = \"" + anoMesStr + "\"");
            ((IJavaScriptExecutor)driver).ExecuteScript("document.getElementById(\"dadosPesquisaDeclaracoes:dataReferenciaInicioMonthPicker:calendar_input\").value = \"" + anoMesStr + "\"");
            ((IJavaScriptExecutor)driver).ExecuteScript("document.getElementById(\"dadosPesquisaDeclaracoes:dataEntregaInicio:calendar_input\").value = \"" + anoMesDiaStr + "\"");

            //Pesquisar (quando concluir cria um elemento com id=PesquisaConcluida)
            /*((IJavaScriptExecutor)driver).ExecuteScript("PrimeFaces.ab({s:\"dadosPesquisaDeclaracoes:pesquisa\",p:\"dadosPesquisaDeclaracoes\",u:\"dadosPesquisaDeclaracoes listaDeclaracoes\",onst:function(cfg){PF('frawPageBlocker').show(); try{PF('varTabelaDeclaracoes').getPaginator().setPage(0)}catch(err){};},onco:function(xhr,status,args){PF('frawPageBlocker').hide();" +
                "document.body.innerHTML += '<div id=\"PesquisaConcluida\"></div>';}});return false;");
                */

            ButtonRunOnClick(By.Id("dadosPesquisaDeclaracoes:pesquisa"));

            //Esperar que acabe de pesquisar (espera que apareça um elemento com id=PesquisaConcluida)
            Thread.Sleep(200);
            var waitableDriver = new WebDriverWait(driver, TimeSpan.FromSeconds(10));
            var element = waitableDriver.Until(ExpectedConditions.InvisibilityOfElementLocated(By.Id("j_idt28_blocker")));


            if (Util.IsElementPresent(driver, By.ClassName("ui-datatable-empty-message")))
            {
                //Se a procura não encontrou nenhuma declaração
            }
            else
            {
                //Se a procura encontrou alguma declaração
                var numExtratos = driver.FindElements(By.XPath("//*[@id=\"formListaDeclaracoes:tabelaDeclaracoes_data\"]/*")).Count;
                for (int i = 0; i < numExtratos; i++)
                {
                    ExpectDownload();
                    //Obtem o codigo que o botão de imprimir tem e executa-o, transferindo assim o PDF
                    ButtonRunOnClick(By.Id("formListaDeclaracoes:tabelaDeclaracoes:" + i + ":imprimirExtrato"));

                    WaitForDownloadFinish(GenNovoNomeFicheiro(Definicoes.estruturaNomesFicheiros.SS_ExtratoRemun),
                        Declaracao.SS_ExtratoRemun, mes);
                    Thread.Sleep(1000);
                }
            }
        }

        internal static void SSPedirCertidao(int ano)
        {
            driver.Navigate().GoToUrl("https://app.seg-social.pt/ssd/pedido_dsc.aspx");
            Thread.Sleep(1000);
            ClickButtonWaitForItToAppear(By.Id("ctl00_ContentPlaceHolder1_btn_iniciar"));

            driver.FindElement(By.Id("ctl00_ContentPlaceHolder1_txt_resp")).SendKeys(empresaAutenticada.NomeDoResponsavel);
            driver.FindElement(By.Id("ctl00_ContentPlaceHolder1_txt_telefone")).SendKeys(empresaAutenticada.TelefoneDoResponsavel);
            driver.FindElement(By.Id("ctl00_ContentPlaceHolder1_txt_mail")).SendKeys(empresaAutenticada.EmailDoResponsavel);
            driver.FindElement(By.Id("ctl00_ContentPlaceHolder1_chk_4")).Click();
            Thread.Sleep(500);
            driver.FindElement(By.Id("ctl00_ContentPlaceHolder1_btn_enviar")).Click();
        }

        internal static void SSDownloadUltimaCertidao(int ano)
        {
            driver.Navigate().GoToUrl("https://app.seg-social.pt/ssd/listadsc.aspx");
            Thread.Sleep(2000);
            string estadoSituacao = driver.FindElement(By.XPath("//*[@id=\"ctl00_ContentPlaceHolder1_MyDataGrid\"]/tbody/tr[2]/td[4]")).Text;
            if (estadoSituacao != "Regularizada")
                throw new Exception("Situação não regularizada: " + estadoSituacao);

            ExpectDownload();

            string data = driver.FindElement(By.XPath("//*[@id=\"ctl00_ContentPlaceHolder1_MyDataGrid\"]/tbody/tr[2]/td[3]")).Text;
            string[] splitData = data.Split('-');
            object fileNameParametros = new { ano = splitData[2], mes = splitData[1], dia = splitData[0] };

            //Obtem o codigo do documento para o transferir
            string numeroSequencial = driver.FindElement(By.XPath("//*[@id=\"ctl00_ContentPlaceHolder1_MyDataGrid\"]/tbody/tr[2]/td[1]")).Text;
            ((IJavaScriptExecutor)driver).ExecuteScript("window.open('doc_dsc.aspx?doc=" + numeroSequencial + "')");

            WaitForDownloadFinish(GenNovoNomeFicheiro(Definicoes.estruturaNomesFicheiros.SS_Transferir_Ultima_Certidao, fileNameParametros),
                Declaracao.SS_Transferir_Ultima_Certidao, 0);
            Thread.Sleep(1000);
            
        }
    }
}

File: novo/downloader_ss.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time

class DownloaderSS:
    @staticmethod
    def download_extrato_remuneracoes(driver, ano, mes, expect_download, wait_for_download_finish, gen_novo_nome_ficheiro, declaracao, estrutura_nomes_ficheiros):
        driver.get("https://app.seg-social.pt/ptss/gr/pesquisa/consultarDR?dswid=7064&frawMenu=1")
        ano_mes_str = f"{ano}-{'%02d' % mes}"
        ano_mes_dia_str = f"{ano}-{'%02d' % mes}-01"
        driver.execute_script(f"document.getElementById('dadosPesquisaDeclaracoes:dataReferenciaFimMonthPicker:calendar_input').value = '{ano_mes_str}'")
        driver.execute_script(f"document.getElementById('dadosPesquisaDeclaracoes:dataReferenciaInicioMonthPicker:calendar_input').value = '{ano_mes_str}'")
        driver.execute_script(f"document.getElementById('dadosPesquisaDeclaracoes:dataEntregaInicio:calendar_input').value = '{ano_mes_dia_str}'")
        DownloaderSS.button_run_onclick(driver, (By.ID, "dadosPesquisaDeclaracoes:pesquisa"))
        time.sleep(0.2)
        WebDriverWait(driver, 10).until(EC.invisibility_of_element_located((By.ID, "j_idt28_blocker")))
        if driver.find_elements(By.CLASS_NAME, "ui-datatable-empty-message"):
            return  # No declarations found
        num_extratos = len(driver.find_elements(By.XPATH, "//*[@id='formListaDeclaracoes:tabelaDeclaracoes_data']/*"))
        for i in range(num_extratos):
            expect_download()
            DownloaderSS.button_run_onclick(driver, (By.ID, f"formListaDeclaracoes:tabelaDeclaracoes:{i}:imprimirExtrato"))
            wait_for_download_finish(gen_novo_nome_ficheiro(estrutura_nomes_ficheiros.SS_ExtratoRemun), declaracao.SS_ExtratoRemun, mes)
            time.sleep(1)

    @staticmethod
    def button_run_onclick(driver, by):
        element = driver.find_element(*by)
        on_click_code = element.get_attribute('onclick')
        if on_click_code:
            driver.execute_script(on_click_code)
        else:
            href = element.get_attribute('href')
            driver.get(href)
